F821 false-positive for SQLAlchemy mappings with not yet declared classes

The described behavior happens since 0.3.3. It might be related to https://github.com/astral-sh/ruff/issues/10340

It is triggered throughout my SQLAlchemy models in cases where a model is not defined yet while used in  a type annotation

Minimal repro:

```python
from __future__ import annotations

from sqlalchemy.orm import DeclarativeBase, Mapped


class Base(DeclarativeBase):
    some_mapping: Mapped[list[Bar]] | None = None # F821 Undefined name `Bar`
    simplified: list[Bar] | None = None # F821 Undefined name `Bar`


class Bar:
    pass
```

I can only trigger the error when inheriting from `DeclarativeBase`. Without this inheritance or when inheriting from another class, things are fine.

We are also hitting this issue in this project: https://github.com/arista-netdevops-community/anta

If you install the project locally with 0.3.3 you should see a bunch of F821 errors:

```bash
(.anta-dev) ~/git_projects/anta (main ✗) ruff check 
anta/models.py:273:27: F821 Undefined name `ResultOverwrite`
anta/models.py:274:18: F821 Undefined name `Filters`
anta/tests/bfd.py:40:25: F821 Undefined name `BFDPeer`
anta/tests/bfd.py:93:25: F821 Undefined name `BFDPeer`
anta/tests/connectivity.py:35:21: F821 Undefined name `Host`
anta/tests/connectivity.py:94:25: F821 Undefined name `Neighbor`
anta/tests/interfaces.py:176:26: F821 Undefined name `InterfaceState`
anta/tests/interfaces.py:535:26: F821 Undefined name `InterfaceDetail`
anta/tests/routing/bgp.py:175:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:285:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:397:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:509:25: F821 Undefined name `BgpNeighbor`
anta/tests/routing/bgp.py:581:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:653:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:719:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:785:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:846:31: F821 Undefined name `VxlanEndpoint`
anta/tests/routing/bgp.py:909:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:968:25: F821 Undefined name `BgpPeer`
anta/tests/security.py:303:28: F821 Undefined name `APISSLCertificate`
anta/tests/security.py:462:33: F821 Undefined name `IPv4ACL`
anta/tests/security.py:471:27: F821 Undefined name `IPv4ACLEntry`
anta/tests/services.py:106:27: F821 Undefined name `DnsServer`
anta/tests/services.py:159:23: F821 Undefined name `ErrDisableReason`
Found 24 errors.
```


Thanks

--- a/crates/ruff_linter/resources/test/fixtures/pyflakes/F821_27.py
+++ b/crates/ruff_linter/resources/test/fixtures/pyflakes/F821_27.py
@@ -33,3 +33,16 @@ class MyClass:
 baz: MyClass
 eggs = baz  # Still invalid even when `__future__.annotations` are enabled
 eggs = "baz"  # always okay
+
+# Forward references:
+MaybeDStr: TypeAlias = Optional[DStr]  # Still invalid even when `__future__.annotations` are enabled
+MaybeDStr2: TypeAlias = Optional["DStr"]  # always okay
+DStr: TypeAlias = Union[D, str]  # Still invalid even when `__future__.annotations` are enabled
+DStr2: TypeAlias = Union["D", str]  # always okay
+
+class D: ...
+
+# More circular references
+class Leaf: ...
+class Tree(list[Tree | Leaf]): ...  # Still invalid even when `__future__.annotations` are enabled
+class Tree2(list["Tree | Leaf"]): ...  # always okay

--- a/crates/ruff_linter/src/checkers/ast/mod.rs
+++ b/crates/ruff_linter/src/checkers/ast/mod.rs
@@ -937,6 +937,7 @@ impl<'a> Visitor<'a> for Checker<'a> {
             && !self.semantic.in_deferred_type_definition()
             && self.semantic.in_type_definition()
             && self.semantic.future_annotations()
+            && (self.semantic.in_annotation() || self.source_type.is_stub())
         {
             if let Expr::StringLiteral(ast::ExprStringLiteral { value, .. }) = expr {
                 self.visit.string_type_definitions.push((

